### DISCUSSION TITLE
Re-reading a trace after it finished

### DISCUSSION THREAD

**Discussion by alberto-ros** [ChampSim/ChampSim#337] [Ideas]
Hi everyone, 

Currently ChampSim executes the number of instructions that it takes by parameter. In case the trace is shorter than the number of instructions specified, ChampSim re-loads it. I wonder if this is the desired behavior for the simulator. For example, if I warm up 50M instruction and execute 50M instruction, a prefetcher could learn the whole trace pattern on the warm up and repeated it in the execution in a very successful way. 

In many of the my ChampSim version I deactivated the trace re-load, but I always need to port it again in the next version, so I guess it would be appreciated to have it here as a feature. Perhaps to make it flexible we can use the 0 as number of instructions to run the simulation just until the end of the trace. 

What do you think?

Alberto

**Comment by ngober** [comment]
I like this idea, and you're not the first person I've heard mention this. What if there were a command line option like `--repeat=[always,never,error]`. We could also make an unspecified (rather than specified 0) number of simulation instructions to mean "entire trace"

**Comment by alberto-ros** [comment]
I like the idea of if not specifying --simulation_instructions then it will run until the end without repeating. Makes to me more sense than having a default value for simulations_instructions if not specified.

**Comment by ngober** [comment]
If we specify the number of simulation instructions, but reach the end of the trace before then, should that be an error, or should we terminate early?

**Comment by alberto-ros** [comment]
If we do not want to change much the current behavior, I would say to re-load the trace, as it happens now. The only change would be when the number is not specified. So old commands would do exactly the same and we get backwards compatibility :-)

**Comment by ngober** [comment]
I opened #342 to address this.
